backend/dao/winesDAO.js

```diff
@@ -1,3 +1,305 @@
+        query.price_eur = { $gte: 0, $lt: 20.0 };
+      } else if (filters.price_eur == "med") {
+        query.price_eur = { $gte: 20.0, $lt: 50.0 };
+      } else if (filters.price_eur == "high") {
+        query.price_eur = { $gte: 50, $lt: 75 };
+      } else if (filters.price_eur == "exp") {
+        query.price_eur = { $gte: 75 };
+      }
+    } else {
+      for (let i = 0; i < filters.price_eur.length; i++) {
+        if (filters.price_eur[i] == "low") {
+          singleFilter = { $gte: 0, $lt: 20.0 };
+        } else if (filters.price_eur[i] == "med") {
+          singleFilter = { $gte: 20.0, $lt: 50.0 };
+        } else if (filters.price_eur[i] == "high") {
+          singleFilter = { $gte: 50, $lt: 75 };
+        } else if (filters.price_eur[i] == "exp") {
+          singleFilter = { $gte: 75 };
+        }
+        filterObject = { price_eur: singleFilter };
+        //make it a list to prepare for query structure Mongodb accepts
+        filter_.push(filterObject);
+        console.log("filterObject: ", filterObject);
+        console.log("filter list: ", filter_);
+      }
+      console.log("Before query assignment: ", query);
+      //query = { $or: filter_ };
+      //query.$and = [{ $or: filter_ }];
+
+      query1 = { $or: filter_ };
+      price_type.push(query1);
+      query.$and = price_type;
+      /*       console.log("price_type: ", price_type);
+      console.log("after query assignment: ", query);
+      console.log("+++++++++++++++++++++++++");
+      console.log("+++++++++++++++++++++++++"); */
+    }
+  }
+  //filter = {};
+  filter_ = [];
+  singleFilter = {};
+  return query;
+}
+function foodSelectors(filters) {
+  if ("food_names" in filters) {
+    query.food_names = { $eq: filters["food_names"] };
+
+    return query;
+  }
+}
+function veganSelectors(filters) {
+  if (filters.vegan == "true") {
+    query.vegan = { $eq: true };
+  } else if (filters.vegan == "false") {
+    query.vegan = { $eq: false };
+  } else {
+    console.log("Invalid query");
+  }
+
+  return query;
+}
+//* *******************************
+//* ***Function profileSelectors***
+//* *******************************
+//? function works: to detect if both sweet and dry are provided, all wines will return (as sweet and wine are on same scale).
+//? splicing the sweet and dry from params incase bitter is also there, so that then bitter can be passed through alone.
+//? if only one param, (type is string), take the single param and query it
+//? if multiple params (type is object), create a list and return query of multiple parameters
+//* *******************************
+//* *******************************
+//* *******************************
+
+function profileSelectors(filters) {
+  //let filter_ = [];
+
+  if (filters.profile.includes("dry") && filters.profile.includes("sweet")) {
+    do {
+      for (var i = filters.profile.length - 1; i >= 0; i--) {
+        if (filters.profile[i] === "sweet") {
+          filters.profile.splice(i, 1);
```

Comment:
One thing that relates to my previous comment about shared state.
We shouldn't mutate objects passed as parameter, this is very prone to bugs, because if we pass an object to a function call in any piece of code, we naturally don't expect that it will get mutated by the function.
Also if we have functions that mutate the parameters, we would need to read every single function being called in our code to understand what is the shape of an object after a function call that it is passed as parameter.

There are exceptions to this of course, but it's usually predefined patterns, like the express app we mutate the response object, sort of actually, we call methods in the response parameter to shape response we want to a certain request.